### PR TITLE
Standalone PR sets additional Jira fields

### LINKED ISSUE #102: Standalone PR sets additional Jira fields
We need a mechanism that can set additional fields to an issue when issue is created. Something like
```
additional-fields: {customfield_007: true, customfield_42: 'painful'}
```

These should be added to the payload of creating issue, as they are.

This can be used for customization (like setting a different issue type), or prefilling mandatory fields of a project.

Issue creation for a field schema with required fields returns
```
{
  "errorMessages": [],
  "errors": {
    "versions": "Affects versions is required."
  }
}
```

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #102

### CODE REVIEW

File: PullRequestCreated/ReadMe.md [hunk 1]
```diff
@@ -25,6 +25,16 @@ Token to access the Jira API.
 
 Jira project key that is used to create new issues.
 
+### `additional-fields`
+
+Additional fields to set when creating the Jira issue. 
+
+```yaml
+additional-fields: '{ "fixVersions": [ { "id": "15601" } ] }'
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:33]: FixVersion is a short live value that changes often, therefore it's not very suitable for the demo here.  Can you use something else, like label, component, or priority? Something that can be done in a simple way.

File: PullRequestCreated/PullRequestCreated.ts [hunk 2]
```diff
@@ -38,6 +38,23 @@ class PullRequestCreated extends OctokitAction {
     }
   }
 
+  private additionalFields(parameters: IssueParameters): any {
+    let addutinalFields = {};
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:46]: metabuerguer 😆 
```suggestion
    let additionalFields = {};
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 3]
```diff
@@ -38,6 +38,23 @@ class PullRequestCreated extends OctokitAction {
...
+    if (parameters?.parent) {
+      addutinalFields = { parent: { key: parameters.parent } };
+    }
+    const inputAdditionFields = this.getInput('additional-fields');
+    if (inputAdditionFields) {
+      try {
+        const fields = JSON.parse(inputAdditionFields);
+        addutinalFields = { ...addutinalFields, ...fields };
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:54]: This should not be inside the `try`. What about extracting `parseAdditionalFields` that would return the object, or null? And here, we would use the resulting value
```suggestion
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 4]
```diff
@@ -42,6 +42,23 @@ class PullRequestCreated extends OctokitAction {
     }
   }
 
+  private additionalFields(parameters: IssueParameters): any {
+    let addutinalFields = {};
+    if (parameters?.parent) {
+      addutinalFields = { parent: { key: parameters.parent } };
+    }
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:49]: This looks weird. What about removing `issueType: string` from `createIssue`, make `additionalFields` mandatory and pass issueType via that?
Then here, there would be something like 
```
  return { ...this.parseAdditionalFields(), ...parameters}
```

File: PullRequestCreated/PullRequestCreated.ts (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply PullRequestCreated/PullRequestCreated.ts:54]: ..or empty object `{ }` instead of null, after updatin the other comment about `...` usage

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Let's simplify this

File: PullRequestCreated/PullRequestCreated.ts (hunk 2 above)
Comment (sebastien-marichal) [reply PullRequestCreated/PullRequestCreated.ts:46]: The heat is not helping me today... 🤦

File: PullRequestCreated/ReadMe.md [hunk 5]
```diff
@@ -25,6 +25,16 @@ Token to access the Jira API.
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:25]: Please update the intro to document the functionality that this action does. Like updating issue title and issue description

File: PullRequestCreated/ReadMe.md [hunk 6]
```diff
@@ -7,6 +7,10 @@ This action will take the first non-`Sub-task` ticket and use it as a parent for
 - if the parent is not an `Epic`, the created ticket will be a `Sub-task`
 - if there is no parent, the created ticket will be a `Task`
 
+This action will also update the PR title and description to mention the right Jira tickets.
+The action will link all mentioned tickets in the title into the description.
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:11]:
```suggestion
The action will update the PR description with links to all mentioned Jira tickets for easy navigation.
```

File: PullRequestCreated/ReadMe.md [hunk 7]
```diff
@@ -7,6 +7,10 @@ This action will take the first non-`Sub-task` ticket and use it as a parent for
 - if the parent is not an `Epic`, the created ticket will be a `Sub-task`
 - if there is no parent, the created ticket will be a `Task`
 
+This action will also update the PR title and description to mention the right Jira tickets.
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:10]: I don't understand what this means. It looks like a combination of the next two points.
```suggestion
```

File: PullRequestCreated/ReadMe.md [hunk 8]
```diff
@@ -7,6 +7,10 @@ This action will take the first non-`Sub-task` ticket and use it as a parent for
...
+The action will link all mentioned tickets in the title into the description.
+When creating a ticket, the action will add it to the PR title.
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:12]: This should be before the point about updating the description, to keep them in (chrono)logical order
```suggestion
When creating a new Jira issue, the action will add the new issue ID to the PR title.
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 9]
```diff
@@ -43,20 +43,19 @@ class PullRequestCreated extends OctokitAction {
   }
 
   private additionalFields(parameters: IssueParameters): any {
-    let addutinalFields = {};
-    if (parameters?.parent) {
-      addutinalFields = { parent: { key: parameters.parent } };
-    }
+    return { ...this.parseAdditionalFields(), ...parameters }
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:46]: Can we embed the `{...}` into the call site? It looks like unneeded indirection now

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, with some final polishing

**Pull request merged**